source/tests/utilities/tcp_echo_server.h

```diff
@@ -0,0 +1,58 @@
+#include <iostream>
+#include <memory>
+#include <cstring>
+#ifdef _WIN32
+#include <winsock2.h>
+#else
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <unistd.h>
+#endif
+
+#ifndef _WIN32
+typedef int SOCKET;
+#endif
+
+class TcpEchoSession : public std::enable_shared_from_this<TcpEchoSession> {
+public:
+    TcpEchoSession(SOCKET socket_fd)
+    : socket_fd_(socket_fd)
+    {
+    }
+
+    ~TcpEchoSession() {
+#ifdef _WIN32
+    closesocket(socket_fd_);
+#else
+    close(socket_fd_);
+#endif
+    }
+
+    void start()
+    {
+        do_read();
```

Comment:
While it makes sense for read to call write, it doesn't make sense for write to call read again. That will blow up the stack after some time. Rather, have it return. Here, just do an infinite loop:
```
for (;;) {
   do_read();
}
```